game-core/src/main/java/games/strategy/triplea/delegate/battle/TargetGroup.java

```diff
@@ -0,0 +1,94 @@
+package games.strategy.triplea.delegate.battle;
+
+import com.google.common.collect.Sets;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.Getter;
+import lombok.Setter;
+import org.triplea.java.collections.CollectionUtils;
+
+/** Group of firing units and their targets. */
+@Getter
+@Setter
+public class TargetGroup {
+
+  private final Set<UnitType> firingUnitTypes;
+  private final Set<UnitType> targetUnitTypes;
+
+  public TargetGroup(final UnitType firingUnitType, final Set<UnitType> targetUnitTypes) {
+    firingUnitTypes = Sets.newHashSet(firingUnitType);
+    this.targetUnitTypes = targetUnitTypes;
+  }
+
+  public Collection<Unit> getFiringUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(firingUnitTypes));
+  }
+
+  public Collection<Unit> getTargetUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(targetUnitTypes));
+  }
+
+  public String getMessage(final GamePlayer player) {
+    return player.getName() + " fire, ";
+  }
+
+  /**
+   * Find all target groupings for firing units and enemy units based on canNotTarget and
+   * canNotBeTargetedBy attributes. Also takes into account if any destroyers are present which
+   * cancel canNotBeTargetedBy. Sort all the target groups so most restrictive appear first.
+   */
+  public static List<TargetGroup> newTargetGroups(
```

Comment:
A test of this method would help review and add confidence this is completely working as expected. Such a test would give design feedback as well and potentially help drive further break down in responsibility to make it easier to test and understand.

For example, extracting 'findTargets' to its own module would be a first simplification that would help decompose the problem, ie: a targetFinder that is perhaps injected into this class.

Reply:
Refactored this a bit to break it up and rename some variables. I would eventually like to add unit tests to this since it is a pretty isolated functionality that should be easy to test. Not sure I'll have the time to do that in the next few days though so might be a better follow up.

Reply:
Add unit tests.